contents/blog/how-we-improved-feature-flags-resiliency.md

```diff
@@ -0,0 +1,145 @@
+
+Overall, this has been working pretty well. There's growing pains of replicating every new feature in flags in all server-side SDKs, but other than that the core functionality is solid.
+
+
+## Making feature flags reliable
+
+As much as I'd love to always be on the happy path, the laws of thermodynamics disagree and things tend to derail. We want feature flags to be reliable, so that when even our database goes down, feature flags can still work fine.
+
+It's also worth noting that there's no such thing as reliable-no-matter-what-happens. An asteroid wiping out datacenters across the world, an AGI taking over, or someone hacking into PostHog to bring it down are all very unlikely, but if they do occur there's not much we can do.
+
+Thankfully, asteroids don't hit us every week. Pgbouncer issues though are indeed a weekly annoyance.
+
+So, when thinking about reliability, we want to prioritise defending against things that happen frequently, or have a high chance of occurring over time. This includes things like redis, postgres, or pgbouncer going down. Then, if we have the resources and nothing better to prioritise, we can focus on defending against asteroids.
+
+Today, we can't yet defend against asteroids, nor the entire infrastructure going down, but for other things, like postgres, we've found ways to defend against this, leveraging our special problem constraints.
+
+### Partial flag evaluation
+
+Since flags are evaluated multiple times in a session, sending a partial response when we can't access the database is much more preferable to retrying and not sending a response at all. Further, if a client is waiting for flag evaluation before loading their content, we do not want to slow this down. We want to return results as soon as possible.
+
+So, we enable partial updates on our client side SDKs. Whenever there's an error computing all flags, we do a partial update: keep the old values for flags we failed to compute, and use the new value for flags we didn't fail to compute.
+
+As we've noted before, the only flags that can fail evaluation are ones that depend on specific properties. Further, if a property change is triggering flag evaluation, the client SDK can send these new properties alongside the request, and we use these properties as overrides for flag evaluation.
+
+TK: rephrase - Specialness: Flags that affect the most people will almost never go down. Flags affecting a small % (property based) can be unavailable more often.
+
+
+Since the server-side SDKs are stateless, this partial evaluation model doesn't really work. Local evaluation is the best way to maintain reliability, or sending known properties alongside requests to make evaluation of property based flags reliable.
+
+
+### Flag evaluation when databases are down
```

Comment:
Doesn't the paragraph about partial flag evaluation also fall under this? Of what happens when the database can't be accessed or is down

Reply:
then that would probably make it clearer to distinguish it between what the client does vs server more clearly upfront?

Reply:
yeah makes sense, did that